src/lib/markbind/src/preprocessors/variablePreprocessor.js

```diff
@@ -0,0 +1,468 @@
+const cheerio = require('cheerio');
+const fs = require('fs');
+const path = require('path');
+
+const _ = {};
+_.clone = require('lodash/clone');
+_.cloneDeep = require('lodash/cloneDeep');
+_.has = require('lodash/has');
+_.isArray = require('lodash/isArray');
+_.isEmpty = require('lodash/isEmpty');
+
+const md = require('../lib/markdown-it');
+const utils = require('../utils');
+const njUtil = require('../utils/nunjuckUtils');
+const urlUtils = require('../utils/urls');
+
+const {
+  ATTRIB_CWF,
+  IMPORTED_VARIABLE_PREFIX,
+} = require('../constants');
+
+
+/**
+ * Variable preprocessor to recursively extract and render pages' variables.
+ * Houses a site-wide variable cache.
+ */
+class VariablePreprocessor {
+  constructor(rootPath, baseUrlMap) {
+    /**
+     * Root site path
+     * @type {string}
+     */
+    this.rootPath = rootPath;
+
+    /**
+     * Set of sites' root paths
+     * @type {Set<string>}
+     */
+    this.baseUrlMap = baseUrlMap;
+
+    /**
+     * Map of sites to their variables
+     * @type {Object<string, any>}
+     */
+    this.userDefinedVariablesMap = {};
+
+    /*
+     * Page level
+     */
+
+    /**
+     * Map of files to another object containing its variable names matched to values
+     * @type {Object<string, Object<string, string>>}
+     */
+    this.variableLookup = {};
```

Comment:
these 3 variables are the static variables originally from `Parser`

I changed these from `Map<string, Map<string, string>>` to `Object<...>` for the following reasons:
- familiarity - objects are being used mostly everywhere else in markbind, if not all other places.
- order matters, but not in the way of retrieval (get/set)
  - when we extract variables are sensitive to declaration order - this is still done the same way as before, which is the order that say `$('variable')` iterates through the elements. This is also the only order that matters
  - when we retrieve variables, briefly ( checked by scrutinizing the usages of `variableLookup/fileAliasesMap` ), order does not matter, because variable names are already unique after extraction
- the key value pairs are all primitives that don't require use of a `Map` ( performance )